app/classifier/tasks/survey/chooser/Choices.jsx

```diff
@@ -0,0 +1,183 @@
+    this.sortChoiceButtons();
+  }
+
+  componentDidUpdate() {
+    this.sortChoiceButtons();
+  }
+
+  sortChoiceButtons() {
+    // overrides default DOM focus order by sorting the buttons according to task.choicesOrder
+    const newChoiceButtons = [];
+    this.choiceButtons
+      .filter(Boolean)
+      .map((button) => {
+        const choiceID = button.getAttribute('data-choiceID');
+        const index = this.props.task.choicesOrder.indexOf(choiceID);
+        newChoiceButtons[index] = button;
+      });
+    this.choiceButtons = newChoiceButtons.filter(Boolean);
+  }
+
+  howManyColumns({ length }) {
+    if (length <= 5) {
+      return 1;
+    } else if (length <= 20) {
+      return 2;
+    } else {
+      return 3;
+    }
+  }
+
+  whatSizeThumbnails({ length }) {
+    if (length <= 5) {
+      return 'large';
+    } else if (length <= 10) {
+      return 'medium';
+    } else if (length <= 20) {
+      return 'small';
+    } else if (this.props.task.alwaysShowThumbnails) {
+      return 'small';
+    } else {
+      return 'none';
+    }
+  }
+
+  handleKeyDown(choiceId, e) {
+    const index = this.choiceButtons.indexOf(document.activeElement);
+    let newIndex;
+    switch (e.which) {
+      case BACKSPACE:
+        this.props.onRemove(choiceId);
+        e.preventDefault();
+        break;
+      case UP:
+        newIndex = index - 1;
+        if (newIndex === -1) {
+          newIndex = this.choiceButtons.length - 1;
+        }
+        this.choiceButtons[newIndex].focus();
+        e.preventDefault();
+        break;
+      case DOWN:
+        newIndex = (index + 1) % this.choiceButtons.length;
+        this.choiceButtons[newIndex].focus();
+        e.preventDefault();
+        break;
+      default:
+    }
+  }
+
+  render() {
+    const { filteredChoices, task, translation } = this.props;
+    this.choiceButtons = [];
+    const selectedChoices = this.props.annotation.value.map(item => item.choice);
+    const columnsCount = this.howManyColumns(filteredChoices);
+    const sortedFilteredChoices = sortIntoColumns(filteredChoices, columnsCount);
+    const thumbnailSize = this.whatSizeThumbnails(sortedFilteredChoices);
+    const choiceNotPresent = this.props.focusedChoice.length === 0 || 
+      filteredChoices.indexOf(this.props.focusedChoice) === -1;
+    return (
+      <div className="survey-task-chooser-choices" data-thumbnail-size={thumbnailSize} data-columns={columnsCount}>
+        {sortedFilteredChoices.length === 0 && <div><em>No matches.</em></div>}
+        {sortedFilteredChoices.map((choiceId, i) => {
+          const choice = task.choices[choiceId];
+          const chosenAlready = selectedChoices.indexOf(choiceId) > -1;
+          let tabIndex = -1;
+          if (i === 0 && choiceNotPresent) {
+            tabIndex = 0;
+          }
+          if (choiceId === this.props.focusedChoice) {
+            tabIndex = 0;
+          }
+          const src = this.props.task.images[choice.images[0]];
```

Comment:
Should this have a default value too?

Reply:
The default value here should be `undefined`, if you don’t want the `Thumbnail` component to generate an img tag.